src/PlaneWaveBasis.jl

```diff
@@ -568,3 +568,35 @@ function gather_kpts(data::AbstractArray, basis::PlaneWaveBasis)
         nothing
     end
 end
+
+# Packing routines used in direct_minimization and newton algorithms.
+# They pack / unpack sets of ψ's (or compatible arrays, such as hamiltonian
+# applies and gradients) to make them compatible to be used in algorithms
+# from IterativeSolvers.
+# Some care is needed here : some operators (for instance K in newton.jl)
+# are real-linear but not complex-linear. To overcome this difficulty, instead of
+# seeing them as operators from C^N to C^N, we see them as
+# operators from R^2N to R^2N. In practice, this is done with the
+# reinterpret function from julia.
+# /!\ pack_ψ does not share memory while unpack_ψ does
+function pack_ψ(basis::PlaneWaveBasis{T}, ψ) where T
+    # TODO as an optimization, do that lazily? See LazyArrays
+    vcat([vec(reinterpret(T, ψk)) for ψk in ψ]...)
+end
+function unpack_ψ(basis::PlaneWaveBasis{T}, x) where T
+
+    model = basis.model
+    filled_occ = filled_occupation(model)
+    n_spin = basis.model.n_spin_components
+    n_bands = div(div(model.n_electrons, filled_occ), n_spin)
+
+    lengths = length.(G_vectors.(basis.kpoints)) .* n_bands
+    ends = cumsum(lengths)
+    # We unsafe_wrap the resulting array to avoid a complicated type for ψ.
+    # The resulting array is valid as long as the original x is still in live memory.
+    x = reinterpret(Complex{T}, x)
+    [unsafe_wrap(Array{Complex{T}},
+                 pointer(@views x[ends[ik]-lengths[ik]+1:ends[ik]]),
+                 (length(G_vectors(basis.kpoints[ik])), n_bands))
+     for ik = 1:length(basis.kpoints)]
```

Comment:
I think this is easier to read if you rewrite it as a `map(1:length(basis.kpoints)) do ik ... end` expression.